### ISSUE TITLE
关于padlde编译执行的问题

### ISSUE THREAD

**Issue by adaxiadaxi** [PaddlePaddle/Paddle#47131]
### 请提出你的问题 Please ask your question

请教下paddle同事，用下图图1这种编译和执行方式，编译出来的程序应该是和输入的program逻辑严格一致的吧，会出现额外添加一些类似于dropout等之类的动作嘛？
![63e3f5565ab2087f33c6327299f6992](https://user-images.githubusercontent.com/58928121/196588872-14f7faa1-a3b3-439f-b2fd-9ce60cae4188.png)

**Comment by GuoxiaWang** [comment]
@adaxiadaxi 是的，理论上，输入的图和编译后的图，执行结果应该是等价的。

**Comment by adaxiadaxi** [comment]
@GuoxiaWang 了解，确实是，如果输入和编译后的图存在不一样，那确实属于大bug，应该早早就被发现和修复的。

请问值班同学，这几个优化器的基类有没有什么区别？我们对接分布式的时候，用到了这几个基类，需要看下要不要都支持适配一下
![67fbed8053a01f418bc190f03b6fd65](https://user-images.githubusercontent.com/58928121/197965450-9c6d9305-545b-49ae-8628-e00f48392e75.png)

**Comment by GuoxiaWang** [comment]
@adaxiadaxi  建议使用 paddle.optimizer

**Comment by adaxiadaxi** [comment]
@GuoxiaWang 另外看这个编译接口，似乎是给分布式专用的，单机的时候似乎没有jit编译program，而是直接run这份program，挨个call算子执行。问下除了GraphCore这种以图对接的形式，其他应该也有厂商是需要jit编译的吧，是在哪里做的？请问值班同学了解这个嘛。
![c06251776404d67c254c2f9e6c900d3](https://user-images.githubusercontent.com/58928121/197967697-41a69b59-0eea-429a-9312-b5e39917578a.png)

**Comment by zhiqiu** [comment]
这个compile主要的作用是将program转为内部的Graph表示，并且可以加上一些pass优化，比如fuse。

**Comment by wanx7130** [comment]
> @adaxiadaxi 建议使用 paddle.optimizer

是不是我可以理解成fluid里的optimize是初始，paddle\optimizer是终章，fluid\contrib类似于是一个中间态？

**Comment by wanx7130** [comment]
> 



> 这个compile主要的作用是将program转为内部的Graph表示，并且可以加上一些pass优化，比如fuse。

是不是可以认为paddle的在所有设备上都是以aot算子来执行的,所以只需要一些图优化动作即可，实际上并不再需要jit的算子编译过程了？

**Comment by qili93** [comment]
> > 
> 
> > 这个compile主要的作用是将program转为内部的Graph表示，并且可以加上一些pass优化，比如fuse。
> 
> 是不是可以认为paddle的在所有设备上都是以aot算子来执行的,所以只需要一些图优化动作即可，实际上并不再需要jit的算子编译过程了？

飞桨静态图在 NV GPU 上的运行流程是将 program 转为飞桨的内部 Graph IR 之后对起进行解析获得算子 IR 再以算子粒度执行 CUDA 或者是 cuDNN 的 AOT 算子。

而类似 Graphcore 或者是 TensorRT 这样图接入的方式，到 Graph IR 之后就会转化为 PopART 或者是 TensorRT 的图对象 （这里也是以算子 IR 进行算子粒度的 OP 转化，但都是 IR 层面的），构建得到 PopART 或者是 TensorRT 的图对象， 之后根据其图引擎的要求进行图编译和图执行。

如果您这里需要接入 JIT 的算子，需要考虑通过编译器后端的方式接入，可以考虑下 https://github.com/PaddlePaddle/CINN 的编译器方案。

**Comment by onecatcn** [comment]
> 

总结得很到位，新写的代码建议使用paddle.optimizer.

**Comment by onecatcn** [comment]
> > @adaxiadaxi 建议使用 paddle.optimizer
> 
> 是不是我可以理解成fluid里的optimize是初始，paddle\optimizer是终章，fluid\contrib类似于是一个中间态？


总结得很到位，那些(fluid 、contrib)都是历史遗留的优化器，新写的代码建议使用paddle.optimizer.

**Comment by phjbest** [comment]
请问这里的动转静方式，只能是针对前向函数添加描述符吗？是否有方式可以将正反向放到一个静态图中呢？https://www.paddlepaddle.org.cn/documentation/docs/zh/practices/jit/image_search_with_jit.html

**Comment by adaxiadaxi** [comment]
> 请问这里的动转静方式，只能是针对前向函数添加描述符吗？是否有方式可以将正反向放到一个静态图中呢？https://www.paddlepaddle.org.cn/documentation/docs/zh/practices/jit/image_search_with_jit.html

@onecatcn 凯哥，可以看下这个新的问题~支持回复下~

**Comment by onecatcn** [comment]
> 请问这里的动转静方式，只能是针对前向函数添加描述符吗？是否有方式可以将正反向放到一个静态图中呢？https://www.paddlepaddle.org.cn/documentation/docs/zh/practices/jit/image_search_with_jit.html
@Aurelius84 请问能否帮忙回复一下

**Comment by Aurelius84** [comment]
> 请问这里的动转静方式，只能是针对前向函数添加描述符吗？是否有方式可以将正反向放到一个静态图中呢？https://www.paddlepaddle.org.cn/documentation/docs/zh/practices/jit/image_search_with_jit.html

只需要针对被装饰的前向函数添加InputSpec描述即可，反向会自动生成。在 train mode下，被装饰的函数会对应两个program：前向和反向program。

但框架内部 PartialProgramLayer也提供了包含前向和反向的完整Program，参考代码：https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/fluid/dygraph/dygraph_to_static/partial_program.py#L212

**Comment by phjbest** [comment]
> > 请问这里的动转静方式，只能是针对前向函数添加描述符吗？是否有方式可以将正反向放到一个静态图中呢？https://www.paddlepaddle.org.cn/documentation/docs/zh/practices/jit/image_search_with_jit.html
> 
> 只需要针对被装饰的前向函数添加InputSpec描述即可，反向会自动生成。在 train mode下，被装饰的函数会对应两个program：前向和反向program。
> 
> 但框架内部 PartialProgramLayer也提供了包含前向和反向的完整Program，参考代码：https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/fluid/dygraph/dygraph_to_static/partial_program.py#L212

![image](https://user-images.githubusercontent.com/27426054/199228279-0809dc76-1d63-4bb8-a32c-ba24f499e1f5.png)
在PartialProgramLayer中看到虽然可以拿到完整的program，请问在执行的时候还是分别是前向和反向两个program执行是吗？

**Comment by phjbest** [comment]
> 

我的需求是想在一个program拿到包括正向和反向的完整的训练图，但是由于loss.backward() 无法放在由 @declarative 或 @paddle.jit.to_static()修饰的函数中，导致前向和反向执行分别在RunProgramOp和RunProgramGradOp中执行，我看到下面的提示中有提到 please use other API to replace 'backward'，请问other API是指什么API呢，需要自己实现反向吗
![image](https://user-images.githubusercontent.com/27426054/199230608-511d79c9-a7e7-4ba5-a897-a8d2daa4c66d.png)

**Comment by Aurelius84** [comment]
有一个方式可以获取到包含前向和反向完整的program，你可以这样组织代码：
```python
class Net(Layer):
     def __init__(self, ....):
         # .....

     @paddle.jit.to_static
     def forward(self, x, lable):
         # ...... 省略具体前向逻辑
         loss = F.cross_entropy(y, lable)   # <----- 把loss计算放在forward里
         return loss    # <----- 返回loss

net = Net()

x = paddle.randn(....)
lable = paddle.randn(.....)
lable.stop_gradient = True

loss = net(x, label)
loss.backward()

program = net.forward.main_program
```

我们之前不建议把 loss计算逻辑放到forward 函数里，因为动转静功能是支持预测部署的，但loss计算一般不属于预测。如果你想获取反向，可以尝试按照上面组织代码

**Comment by Aliceyins-dev** [comment]
交叉熵计算数据容易溢出：
`import numpy as np
import paddle
import pytest
import paddle.fluid as fluid
from paddle.fluid import compiler

paddle.enable_static()
np.random.seed(33)
main_program = paddle.static.Program()
startup_program = paddle.static.Program()
main_program.random_seed = 33
startup_program.random_seed = 33

def test_cross_entropy():
    with paddle.utils.unique_name.guard():
        with paddle.static.program_guard(main_program=main_program, startup_program=startup_program):

            data = paddle.static.data(name='data', shape=[1, 1000], dtype='float32')
            label = paddle.static.data(name='label', shape=[1, 1000], dtype='float32')
            data.stop_gradient = False
            label.stop_gradient = True
            loss = paddle.fluid.layers.cross_entropy(data, label, soft_label=True)
            feed_list = ["data", "label"]
            fetch_list = [loss.name]
            g = paddle.static.gradients(loss, [data, label])
            print(g)
            fetch_list.append(g[0].name)
            print(fetch_list)
            print(main_program)
            exec = paddle.static.Executor(paddle.CPUPlace())
            exec.run(startup_program)
            x = np.load("./softmax_0.tmp_0.npy")
            x2 = np.load("./label_smooth_0.tmp_0.npy")
            print(x.shape)
            print(x2.shape)
            output_cpu = exec.run(main_program, feed={"data": x, "label": x2}, fetch_list=fetch_list)
            print("output num:", len(output_cpu))
            for i in range(len(output_cpu)):
                print("------------")
                print(fetch_list[i], output_cpu[i].reshape(-1))
test_cross_entropy()`

![image](https://user-images.githubusercontent.com/53935495/203326541-d621d576-d26d-4dd1-a979-16170c82f476.png)
同样的数据在pytorch上计算结果为：
![image](https://user-images.githubusercontent.com/53935495/203326840-af9ba656-7ead-4041-9b02-12e51a531e9e.png)
![image](https://user-images.githubusercontent.com/53935495/203327261-ee8100ee-d172-4723-9610-4d048bdf33ef.png)

**Comment by Aliceyins-dev** [comment]
reduce_sum的误差问题：[https://github.com/PaddlePaddle/Paddle/issues/48252](url)

**Comment by onecatcn** [comment]
> 交叉熵计算数据容易溢出： `import numpy as np import paddle import pytest import paddle.fluid as fluid from paddle.fluid import compiler
> 
> paddle.enable_static() np.random.seed(33) main_program = paddle.static.Program() startup_program = paddle.static.Program() main_program.random_seed = 33 startup_program.random_seed = 33
> 
> def test_cross_entropy(): with paddle.utils.unique_name.guard(): with paddle.static.program_guard(main_program=main_program, startup_program=startup_program):
> 
> ```
>         data = paddle.static.data(name='data', shape=[1, 1000], dtype='float32')
>         label = paddle.static.data(name='label', shape=[1, 1000], dtype='float32')
>         data.stop_gradient = False
>         label.stop_gradient = True
>         loss = paddle.fluid.layers.cross_entropy(data, label, soft_label=True)
>         feed_list = ["data", "label"]
>         fetch_list = [loss.name]
>         g = paddle.static.gradients(loss, [data, label])
>         print(g)
>         fetch_list.append(g[0].name)
>         print(fetch_list)
>         print(main_program)
>         exec = paddle.static.Executor(paddle.CPUPlace())
>         exec.run(startup_program)
>         x = np.load("./softmax_0.tmp_0.npy")
>         x2 = np.load("./label_smooth_0.tmp_0.npy")
>         print(x.shape)
>         print(x2.shape)
>         output_cpu = exec.run(main_program, feed={"data": x, "label": x2}, fetch_list=fetch_list)
>         print("output num:", len(output_cpu))
>         for i in range(len(output_cpu)):
>             print("------------")
>             print(fetch_list[i], output_cpu[i].reshape(-1))
> ```
> 
> test_cross_entropy()`
> 
> ![image](https://user-images.githubusercontent.com/53935495/203326541-d621d576-d26d-4dd1-a979-16170c82f476.png) 同样的数据在pytorch上计算结果为： ![image](https://user-images.githubusercontent.com/53935495/203326840-af9ba656-7ead-4041-9b02-12e51a531e9e.png) ![image](https://user-images.githubusercontent.com/53935495/203327261-ee8100ee-d172-4723-9610-4d048bdf33ef.png)

@Aliceyins-dev Alice，辛苦将这个问题提交一个新的issue，方便值班同学安排对应负责同学处理，谢谢

**Comment by Aliceyins-dev** [comment]
已在issue list中找到交叉熵计算容易溢出的相关解答。
